### PR TITLE
DocComment: Parse references to different objects in docstring

### LINKED ISSUE #3920: Docs generator should detect invalid class names in :class:
e.g. the error fixed in https://github.com/coala/coala/pull/3842 should have been detected when the code was uploaded.
difficulty/low

Opened via [gitter](https://gitter.im/coala/coala/?at=58c8ca28dd08b4b859d1b154) by @jayvdb

### PR DESCRIPTION
**Pull request by RJ722**
Parse cross-reference from doc strings so that they can be validated afterwards. This could have help prevented the error fixed in #3842.
### Checklist

- [X] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [X] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

### CODE REVIEW

**Comment by RJ722** [comment]
Now, since we can parse the referred class, it would be simple enough to check whether or not that class exists.

Is this approach feasible @Flap-Py @jayvdb

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 1]
```diff
@@ -120,6 +125,18 @@ def _parse_documentation_with_symbols(self, param_identifiers,
                 retval_desc = line[return_offset:]
                 parsed.append(self.ReturnValue(desc=retval_desc))
 
+            elif ref_class_identifiers and  \
+                    stripped_line.find(ref_class_identifiers[0]) != -1:
```
Comment (Udayan12167) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: What about other ref_class_identifiers in the tuple?

File: tests/bearlib/languages/documentation/DocumentationCommentTest.py [hunk 2]
```diff
@@ -121,6 +122,11 @@ def test_return_values_default(self):
                     self.ReturnValue(desc=' something2 ')]
         self.check_docstring(doc, expected)
 
+    def test_ref_class_default(self):
+        doc = (' :class:`testClass.someThing.anotherThing`. foo bar')
+        expected = [self.ReferenceClass('testClass.someThing.anotherThing')]
+        self.check_docstring(doc, expected)
+
```
Comment (Udayan12167) [inline tests/bearlib/languages/documentation/DocumentationCommentTest.py]: Umm multiple classes is a possibility?

**Comment by Udayan12167** [comment]
ack https://github.com/coala/coala/pull/4009/commits/d773b72cdfef3d48ed91966d98620c5e18ae9cd2

**Comment by RJ722** [comment]
@Udayan12167 

I just have a slight confusion - where should we add the check to find out whether or not is the class valid?
Should it be in the [PyDocStyleBear](https://github.com/coala/coala-bears/blob/9fb36a9d793ffa4953d4430437fed07f851fdbb2/bears/python/PyDocStyleBear.py) ?

Thanks! :smile:

**Comment by userzimmermann** [comment]
@RJ722 Still working on this? :)

**Comment by RJ722** [comment]
Yeah working on it, will positively submit a patch today. :smile:

**Comment by RJ722** [comment]
@userzimmermann where do we check that whether or not the class referred to, exists or not?
Should it be in the [PyDocStyleBear](https://github.com/coala/coala-bears/blob/9fb36a9d793ffa4953d4430437fed07f851fdbb2/bears/python/PyDocStyleBear.py) ?
Thanks

**Comment by RJ722** [comment]
Note to self: Make changes to deal with issue #3920

**Comment by userzimmermann** [comment]
@RJ722 `PyDocStyleBear` only relies on [pydocstyle](https://github.com/PyCQA/pydocstyle), which is not intended for such checks. IMO we need a new bear for this. Also, it's not really a matter of style ;)

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 3]
```diff
@@ -14,6 +14,11 @@ class DocumentationComment:
     Parameter = namedtuple('Parameter', 'name, desc')
     ReturnValue = namedtuple('ReturnValue', 'desc')
     Description = namedtuple('Description', 'desc')
+    ClassReference = namedtuple('ClassReference', 'ref_class_addr')
+    MethodReference = namedtuple('MethodReference', 'ref_method_addr')
+    FunctionReference = namedtuple('FunctionReference', 'ref_func_addr')
+    ObjectReference = namedtuple('ObjectReference', 'ref_obj_addr')
+    ExceptionReference = namedtuple('ExceptionReference', 'ref_exc_addr')
```
Comment (userzimmermann) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: **+** `ModuleReference`, `DataReference`, `ConstRefrence`, `AttributeReference`

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 4]
```diff
@@ -62,17 +67,29 @@ def parse(self):
...
         """
         if self.language == 'python' and self.docstyle == 'default':
             return self._parse_documentation_with_symbols(
-                (':param ', ':'), ':return:')
+                (':param ', ':'), ':return:', {
+                    'class_identifiers': (':class:`', '`'),
+                    'method_identifiers': (':meth:`', '`'),
+                    'func_identifiers': (':func:`', '`'),
+                    'obj_identifiers': (':obj:`', '`'),
+                    'exc_identifiers': (':exc:`', '`')})
```
Comment (userzimmermann) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: **+** `:mod:`, `:data:`, `:const:`, `:attr:`

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 5]
```diff
@@ -62,17 +71,39 @@ def parse(self):
...
+                    'mod_identifiers': ('@mod`', '`'),
+                    'func_identifiers': ('@func`', '`'),
+                    'data_identifiers': ('@data`', '`'),
+                    'constant_identifiers': ('@const`', '`'),
+                    'class_identifiers': ('@class`', '`'),
+                    'method_identifiers': ('@meth`', '`'),
+                    'attr_identifiers': ('@attr`', '`'),
+                    'exc_identifiers': ('@exc`', '`'),
+                    'obj_identifiers': ('@obj`', '`'),
+                    })
```
Comment (RJ722) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: Move ```})``` to the line above

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 6]
```diff
@@ -125,6 +179,106 @@ def _parse_documentation_with_symbols(self, param_identifiers,
...
+                    stripped_line.find(obj_identifiers[0]) != -1:
+                param_offset = line.find(
+                    obj_identifiers[0]) + len(obj_identifiers[0])
+                splitted = line[param_offset:].split(
+                    obj_identifiers[1], 1)
+
+                ref_obj_addr = splitted[0]
+                parsed.append(self.ObjectReference(
+                    ref_obj_addr=ref_obj_addr))
+
```
Comment (userzimmermann) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: All those `elif` blocks above contain the same algorithm. Please generalize :)

**Comment by RJ722** [comment]
Sorry for delay!
I have changed the algorithm for detection, I think we can make final touches, once we are sure that this new algorithm is correct.

I will be pushing all new tests by morning!

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 7]
```diff
@@ -15,6 +15,7 @@ class DocumentationComment:
     ExceptionValue = namedtuple('ExceptionValue', 'name, desc')
     ReturnValue = namedtuple('ReturnValue', 'desc')
     Description = namedtuple('Description', 'desc')
+    Reference = namedtuple('Reference', ['typeRef', 'ref_addr', 'desc'])
```
Comment (RJ722) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: Proper case: Follow either snake_case or CamelCase 😉

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 8]
```diff
@@ -103,6 +123,35 @@ def _parse_documentation_with_symbols(self,
             section is a named tuple of either ``Description``, ``Parameter``,
             ``ExceptionValue`` or ``ReturnValue``.
         """
+
+        def startswith(line, starter):
```
Comment (RJ722) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: @jayvdb and @userzimmermann  What may be a proper location for the function `startswith`?

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 9]
```diff
@@ -116,6 +165,10 @@ def _parse_documentation_with_symbols(self,
 
             stripped_line = line.strip()
 
+            ref_starters = []
+            for key in ref_identifiers:
+                ref_starters.append(ref_identifiers[key][0])
+
```
Comment (RJ722) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: This should be encapsulated in a function IMO!

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 10]
```diff
@@ -103,6 +123,35 @@ def _parse_documentation_with_symbols(self,
...
+            ":func:`" --> function
+            ":data:`" --> data
+            ":constant:`" --> constant
+            ":class:`" --> class
+            ":meth:`" --> method
+            ":attr:`" --> attribute
+            ":exc:`" --> exception
+            ":obj:`" --> object
+            """
+            typeRef = typeRef.strip()
```
Comment (RJ722) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: I have a gut feeling that this function can be optimised, do we need to? and how?

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 11]
```diff
@@ -149,6 +202,18 @@ def _parse_documentation_with_symbols(self,
...
+                typeRef = _extract_typeRef(ref_identifiers, rawtypeRef)
+                splitted = line.split(rawtypeRef, 1)[1].split(
+                    ref_identifiers[typeRef][1])
+                addr = splitted[0].strip()
+                desc = splitted[1].strip()
+                parsed.append(self.Reference(typeRef=typeRef,
+                                             ref_addr=addr, desc=desc))
+
             # These conditions will take care if the parsed section
             # descriptions are not on the same line as that of it's
```
Comment (RJ722) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: Also need to add a condition to check for `ref_addr` and `desc` in the next line!

**Comment by damngamerz** [comment]
DocumentationAPI is under heavy development...Status Blocked till its final revamp...

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 12]
```diff
@@ -113,6 +127,14 @@ def _parse_documentation_with_symbols(self,
             section is a named tuple of either ``Description``, ``Parameter``,
             ``ExceptionValue`` or ``ReturnValue``.
         """
+        def _reference(line, ref_identifiers):
```
Comment (RJ722) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: Is there a more suitable place for this helper function?

I also plan to remove some of the algorithmic redundancy in the other checks by introducing another helper function  - We may then place both of these together.

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 13]
```diff
@@ -208,7 +240,6 @@ def from_metadata(cls, doccomment, docstyle_definition,
 
         This function just assembles the documentation comment
         itself, without the markers and indentation.
-
```
Comment (jayvdb) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: why this change?

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 14]
```diff
@@ -266,7 +297,6 @@ def from_metadata(cls, doccomment, docstyle_definition,
     def assemble(self):
         """
         Assembles parsed documentation to the original documentation.
-
```
Comment (jayvdb) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: why this change?

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 15]
```diff
@@ -5,6 +5,18 @@
 from functools import lru_cache
 
 
+PY_DOC_REF = {
```
Comment (jayvdb) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: these are sphinx restructured text directives/roles.
c.f. http://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 16]
```diff
@@ -80,10 +93,10 @@ def parse(self):
         """
         if self.language == 'python' and self.docstyle == 'default':
             return self._parse_documentation_with_symbols(
-                (':param ', ':'), (':raises ', ':'), ':return:')
+                (':param ', ':'), (':raises ', ':'), ':return:', PY_DOC_REF)
         elif self.language == 'python' and self.docstyle == 'doxygen':
             return self._parse_documentation_with_symbols(
-                ('@param ', ' '), ('@raises ', ' '), '@return ')
+                ('@param ', ' '), ('@raises ', ' '), '@return ', PY_DOC_REF)
```
Comment (jayvdb) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: I expect there will be differences in doxygen.

File: coalib/bearlib/languages/documentation/DocumentationComment.py (hunk 13 above)
Comment (RJ722) [reply coalib/bearlib/languages/documentation/DocumentationComment.py]: I don't know when these changed, amending the commit anyways.

File: coalib/bearlib/languages/documentation/DocumentationComment.py (hunk 16 above)
Comment (RJ722) [reply coalib/bearlib/languages/documentation/DocumentationComment.py]: https://www.stack.nl/~dimitri/doxygen/manual/autolink.html says:

All words in the documentation that correspond to a documented class and contain at least one non-lower case character will automatically be replaced by a link to the page containing the documentation of the class. If you want to prevent that a word that corresponds to a documented class is replaced by a link you should put a % in front of the word. To link to an all lower case symbol, use \ref.

Comment (RJ722) [reply coalib/bearlib/languages/documentation/DocumentationComment.py]: It would be nearly impossible to parse the references using this mechanism. Also, if we do parse the name, I can't think of a way to infer the reference type (`ref_type`) automatically. 

Should we make this shinx only?

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 17]
```diff
@@ -113,6 +127,14 @@ def _parse_documentation_with_symbols(self,
             section is a named tuple of either ``Description``, ``Parameter``,
             ``ExceptionValue`` or ``ReturnValue``.
         """
+        def _reference(line, ref_identifiers):
+            occurences = []
+            for ref_type in ref_identifiers.keys():
+                ref = line.find(ref_identifiers[ref_type][0])
```
Comment (jayvdb) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: what if the same ref type is used twice in the same line?

File: coalib/bearlib/languages/documentation/DocumentationComment.py (hunk 12 above)
Comment (jayvdb) [reply coalib/bearlib/languages/documentation/DocumentationComment.py]: Place them at the module scope with `_` prefix, so they can be tested separately

File: coalib/bearlib/languages/documentation/DocumentationComment.py (hunk 17 above)
Comment (RJ722) [reply coalib/bearlib/languages/documentation/DocumentationComment.py]: Thank You for the tricky catch. Making changes.

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 18]
```diff
@@ -169,6 +207,16 @@ def _parse_documentation_with_symbols(self,
                 retval_desc = line[return_offset:]
                 parsed.append(self.ReturnValue(desc=retval_desc))
 
+            elif _find_references(line, ref_identifiers):
+                occurences = _find_references(line, ref_identifiers)
+                for ref, _ in occurences:
+                    splitted = line.split(
+                        ref_identifiers[ref][0], 1)[1].split(
```
[the remark this answers is not in the document]
Comment (RJ722) [reply coalib/bearlib/languages/documentation/DocumentationComment.py]: Done!

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 19]
```diff
@@ -169,6 +207,16 @@ def _parse_documentation_with_symbols(self,
...
+                        ref_identifiers[ref][1], 1)
```
[the remark this answers is not in the document]
Comment (RJ722) [reply coalib/bearlib/languages/documentation/DocumentationComment.py]: Oops, how come it slipped underneath coala?

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 20]
```diff
@@ -5,6 +5,41 @@
...
+    :param identifiers:
+        A dict mapping the type of references to a tuple of two strings with
+        which the reference might start and end respectively.
+    :return:
+        A list of two-element tuples containing the type of reference (inferred
+        from the given dict) and the corresponding index of occurence in the
+        line, sorted according to the index.
+    """
+    occurences = []
+    for ref_type in identifiers.keys():
```
[the remark this answers is not in the document]
Comment (RJ722) [reply coalib/bearlib/languages/documentation/DocumentationComment.py]: Thanks! Done

File: coalib/bearlib/languages/documentation/DocumentationComment.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (jayvdb) [reply coalib/bearlib/languages/documentation/DocumentationComment.py]: Because pycodestyle is not that good yet.

But of course it is your responsibility to know PEP8 and write accordingly.

**Comment by jayvdb** [comment]
ack https://github.com/coala/coala/pull/4009/commits/14897780d8a72c73a2ebed284e26128e54f37fa6

**Comment by jayvdb** [comment]
@gitmate-bot rebase

**Comment by jayvdb** [comment]
@gitmate-bot ff
